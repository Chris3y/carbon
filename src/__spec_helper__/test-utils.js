const isUpper = char => char.toUpperCase() === char;
const humpToDash = (acc, char) => `${acc}${isUpper(char) ? `-${char.toLowerCase()}` : char}`;

export const toCSSCase = (str) => {
  return str.split('').reduce(humpToDash, '');
};

export const assertStyleMatch = (styleSpec, component) => {
  Object.entries(styleSpec).forEach(([attr, value]) => {
    expect(component).toHaveStyleRule(toCSSCase(attr), value);
  });
};
