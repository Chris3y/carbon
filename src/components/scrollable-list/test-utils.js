import React from 'react';
import ScrollableListItem from './scrollable-list-item.component';

const makeArray = n => [...Array(n).keys()];

const listItemReducer = ({ nonSelectables = [], customSelectables = [] }) => {
  // generate jsx for selectable list items, based on indexes in config
  return (acc, item, index) => {
    if (nonSelectables.includes(item)) return [...acc, <div />];
    if (customSelectables.includes(item)) return [...acc, <div isSelectable />];

    return [...acc, <ScrollableListItem id={ index } />];
  };
};

const renderListItems = (opts) => {
  return makeArray(opts.num).reduce(listItemReducer(opts), []);
};


const keyPress = (code) => {
  const ev = new KeyboardEvent('keydown', { which: code });
  document.dispatchEvent(ev);
};

const keyMap = {
  UpArrow: '38',
  DownArrow: '40',
  RightArrow: '39',
  LeftArrow: '37',
  Enter: '13',
  Tab: '9'
};

const repeat = action => (n = 1) => makeArray(n).forEach(() => action());

const keyboard = Object.keys(keyMap).reduce((acc, key) => {
  acc[`press${key}`] = () => repeat(keyPress(keyMap[key]));
  return acc;
}, {});

const listFrom = wrapper => wrapper.find('ul');
const childrenFrom = node => node.children();

const hoverList = wrapper => (item) => {
  childrenFrom(listFrom(wrapper)).at(item).simulate('mouseover');
};

const simulateEvent = eventName => wrapper => wrapper.simulate(eventName);
const click = simulateEvent('click');

const selectedItemOf = wrapper => wrapper.state().selectedItem;
const isUnique = (val, index, self) => self.indexOf(val) === index;
const isSelectableGiven = nonSelectables => i => !nonSelectables.includes(i);

const selectedItemReducer = method => wrapper => (acc, i) => {
  method(wrapper)(i);
  return [...acc, selectedItemOf(wrapper)];
};

const arraysEqual = (arr1, arr2) => arr1.sort().join(',') === arr2.sort().join(',');

const assertCorrectTraversal = method => expect => ({ num, nonSelectables = [] }) => (wrapper) => {
  const array = makeArray(num);
  const validIndexes = array.filter(isSelectableGiven(nonSelectables));

  const indexesThatWereSelected = array
    .reduce(selectedItemReducer(method)(wrapper), [])
    .filter(isUnique);
  expect(arraysEqual(validIndexes, indexesThatWereSelected)).toBeTruthy();
};

const assertKeyboardTraversal = assertCorrectTraversal(() => keyboard.pressDownArrow)(expect);
const assertHoverTraversal = assertCorrectTraversal(wrapper => hoverList(wrapper))(expect);

export {
  hoverList,
  selectedItemOf,
  childrenFrom,
  renderListItems,
  keyboard,
  assertKeyboardTraversal,
  assertHoverTraversal,
  listFrom,
  click
};
