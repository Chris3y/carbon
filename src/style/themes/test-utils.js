import generatePalette from '../palette';
import atOpacity from '../utils/at-opacity';
import baseColors from '../color-config';

const addHex = base => `#${base}`;

export const palette = {
  ...generatePalette(baseColors),
  atOpacity,
  ...Object.keys(baseColors).reduce((acc, col) => {
    acc[col] = addHex(baseColors[col]);
    return acc;
  }, {})
};
const blackWithOpacity = atOpacity('#000000');

export const baseThemeConfig = {
  colors: {
    brand: palette.brilliantGreen,
    white: '#FFFFFF',
    error: palette.errorRed,
    warning: palette.gold,
    success: palette.brilliantGreenShade(20),
    info: palette.productBlueShade(3),
    text: {
      body: blackWithOpacity(0.9),
      disabled: blackWithOpacity(0.55),
      placeholder: blackWithOpacity(0.3)
    }
  }
};


const isObject = (obj) => {
  return typeof obj === 'object' && obj !== null;
};

export const assertIsSubset = (obj, comparison) => {
  if (!isObject(obj)) {
    // no further nesting, assert that values are equal
    expect(obj).toEqual(comparison);
    return;
  }

  const objKeys = Object.keys(obj);
  const comparisonKeys = Object.keys(comparison);

  objKeys.forEach((key) => {
    // assert that keys are present
    expect(comparisonKeys.includes(key)).toBeTruthy();

    // repeat for nested objects
    assertIsSubset(obj[key], comparison[key]);
  });
};
