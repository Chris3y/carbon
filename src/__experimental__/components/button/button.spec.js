import React from 'react';
import TestRenderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import 'jest-styled-components';
import Icon from 'components/icon';
import Button from '.';
import { assertStyleMatch } from '../../../__spec_helper__/test-utils';
import BaseTheme from '../../../style/themes/base';
import baseTheme from '../../../style/themes/base';

function render(props, renderer = shallow) {
  return renderer(
    <Button { ...props } />
  );
}

describe('Button', () => {
  describe('when setting the "as" prop', () => {
    it('sets the color variant to "secondary" when nothing is passed', () => {

    });

    it('sets the color variant to "primary"', () => {

    });

    it('sets the color variant to "secondary"', () => {

    });

    it('sets the color variant to "tertiary"', () => {

    });

    it('sets the color variant to "destructive"', () => {

    });

    it('sets the color variant to "darkBackground"', () => {

    });
  });

  describe('when the "disabled" prop is passed and "as" is not set', () => {
    it('sets the color variant to "secondary/disabled"', () => {

    });
  });

  describe('when the no props are passed except children', () => {
    // it('"', () => {
    //   const wrapper = render({ children: 'Foo' });
    //   console.log(wrapper.debug());
    //   expect(wrapper
    //     .contains(
    //       <Button
    //         renderAs='secondary'
    //         size='medium'
    //         disabled={ false }
    //         role='button'
    //         onClick={ undefined }
    //         iconPosition=''
    //         medium
    //       >
    //         Foo
    //       </Button>
    //     )).toBeTruthy();

    it('sets the default variants to "secondary" and "medium"', () => {
      // const wrapper = render({ children: 'foo' }, TestRenderer.create).toJSON();
      // console.log(wrapper);
      // assertStyleMatch({
      //   background: 'transparent',
      //   borderColor: baseTheme.colors.primary,
      //   color: baseTheme.colors.primary,
      //   fontSize: '14px',
      //   height: '40px',
      //   paddingLeft: '24px',
      //   paddingRight: '24px'
      // }, wrapper);
      expect(render({ children: 'mock content' }, TestRenderer.create)).toMatchSnapshot();
    });
  });
});
