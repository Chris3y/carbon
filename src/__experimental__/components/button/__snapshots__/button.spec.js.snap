// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button when the no props are passed except children sets the default variants to "secondary" and "medium" 1`] = `
.c1 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  border: 2px solid transparent;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-flex-flow: wrap;
  -ms-flex-flow: wrap;
  flex-flow: wrap;
  font-weight: 600;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  vertical-align: middle;
  background: transparent;
  border-color: #26A826;
  color: #26A826;
  font-size: 14px;
  height: 40px;
  padding-left: 24px;
  padding-right: 24px;
}

.c1:focus {
  outline: solid 3px #FFB500;
}

.c1 + .c0 {
  margin-left: 16px;
}

.c1:hover {
  background: #006300;
  border-color: #006300;
  color: #FFFFFF;
}

.c1 .carbon-icon {
  margin-left: 8px;
  margin-right: -4px;
}

<button
  className="c0 c1"
  disabled={false}
  role="button"
  size="medium"
>
  
  <span>
    <span
      data-element="main-text"
    >
      mock content
    </span>
  </span>
  
</button>
`;
